dgram: initial support for DATAGRAM extension

LPardue opened the pull request:
With this change, we introduce support for the DATAGRAM frame extension
specified in draft-ietf-quic-datagram-01 and draft-schinazi-quic-h3-datagram-04.

Several API methods have been added to the quiche transport library:

* `enable_dgram()` enables the optional QUIC extension for a  connection.
* `set_dgram_send_max_queue_len()` and `set_dgram_recv_max_queue_len()`
  control internal queue sizes (see below).
* `dgram_max_writable_len()` returns the size of the largest DATAGRAM frame
  payload that can be sent on the connection.
* `dgram_send()` buffers the provided payload data in quiche, a subsequent
  call to `send()` constructs a DATAGRAM frame and QUIC packet.
* `dgram_purge_outgoing()` purges buffered payloads matching a predicate.
* `dgram_recv()` extracts payload data buffered in quiche from a prior call
  to `recv()`.

Internally, a send and receive queue are added to quiche for the purposes of
buffering the payload of DATAGRAM frames. By their specification, these frames
are not flow controlled by QUIC but because of quiche's API design we benefit
some ability to enqueue/dequeue application data independent from the
packetization and network loop. The size of queues is configurable in order
for applications to tailor things to their needs.

Several API methods have been added to the quiche HTTP/3 library, which
internally builds on the transport library:

* `dgram_max_writable_len()` returns the size of the largest DATAGRAM frame
  payload that can be sent on the connection.
* `dgram_send()` buffers the payload for a flow ID.
* `poll_dgram()` processes received DATAGRAMs and generates an 
  `quiche::h3::Event::Datagram` if there is something to read
* `dgram_recv()` extracts payload data buffered in quiche

Previously, quiche-client and quiche-server supported HTTP/0.9, HTTP/3 or both
for any given QUIC connection. During TLS handshake, ALPN would select one of
these and then the applications would create HttpConn objects to handle
request/response. Http09Conn simply invokes quiche transport APIs directly.
Http3Conn invokes quiche transport and H3 APIs.

With DATAGRAM there are many more permutations. Rather than boil the ocean, this
commit introduces support for two modes, with the following CLI parameters that
control how datagrams are used:
```
  --dgram-proto PROTO         DATAGRAM application protocol to use.
  --dgram-count COUNT         Number of DATAGRAMs to send.
  --dgram-data DATA           Data to send for certain types of DATAGRAM application protocol.
```

Valid values for `dgram-proto` are `siduck` and `oneway`.

`siduck` works according to the I-D: the endpoints advertise the ALPN siduck,
siduck-00 and if negotiated, send dgram_count number of DATAGRAM frames. The
client by default sends quack, which can be overridden by the dgram-data
parameter. The server simply attempts to respond to every received DATAGRAM with
${value-ack; it ignores dgram-count and dgram-data. The client counts sent
quacks and received quack acks and reports this. No HTTP requests or responses
are made in this mode.

`oneway` layers on top of HTTP/3. Endpoints unilaterally send dgram-count amount
of HTTP/3 DATAGRAMS to each other and no accounting is done. Clients send on
flow ID 0, server on flow ID 1. By default the client sends quack, the server
sends brrr. Meanwhile, the apps send requests/responses in accordance with all
existing CLI parameters.

ghedo reviewed (changes requested):
Sorry for the delay. I didn't quite go through all changes yet, but at a general level I have a few comments/questions:
* I don't think we need to hide this behind features, since there are no additional dependencies needed and the functionality is not enabled by default anyway.
* Do we really need to duplicate the HTTP/3 `poll()` method and events struct? I guess this might have been needed because of the optional features, but as per comment above I don't think this is needed, so we can merge the poll methods and Events structs together IMO.
* Similarly, it seems like it would be better if we could somehow add dgram support to the existing quiche apps rather then add a whole new crate just for that. It'll likely make it easier to integrate the functionality into the existing interop automation if anything.
* I'm not sure I understand the point of the "purge outgoing" API. Or rather, it seems like it could potentially be very expensive to do that, what would be the use case for it?

xanathar commented:
> I'm not sure I understand the point of the "purge outgoing" API. Or rather, it seems like it could potentially be very expensive to do that, what would be the use case for it?

The rationale behind it, is to allow control of outgoing datagrams by application logic in the case of pending datagrams which have been, de facto, invalidated by application logic.

An example could be video streaming: a server can be compressing video and sending frames on datagrams; let's say frame 30 gets lost, and the client sends a NACK to the server. The server reacts by generating a new I-frame (say, frame 36). The datagram queue might still contain datagrams relative to frames [31..=35] which are P/B-frames using frame-30 as a dependency; given this the server might want to purge all the invalidated frames, before adding the new I-frame.

Given how it's implemented, the callback shouldn't take much time (Rust should solve the generic to the closure at compile time, avoiding indirect calls - hopefully); what might have a performance differences are:
- packet inspection, but that is done by the application, so it's safe to assume that the application itself would optimize for that
- the "packing" of the VecQueue which is contextual to the purging; this is directly proportional to the number of items purged from the queue, and it can be assumed that the cost of moving pointers in a vec is smaller than the sending of an extra, unnecessary, UDP packet (which involves, among other things, a switch to kernel mode sooner or later)

That said, if you think it's an obstacle to PR approval, it's surely an "accessory" part of the whole PR, so it can be removed easily and may be reconsidered later.

LPardue commented:
Thanks for the Feedback @ghedo!

All of these are fair points. @xanathar has answered the purge question and I agree we could split if off if so desired. 

To respond to the remaining points, these are all manifestations of my decision to trying and limit the blast zone of the datagram feature. As you point out, that doesn't *need* to be done and introduces some difficulties of its own. If we think datagram support is a viable candidate to be included in quiche, all of these things can be simplified.

I do have a concern about the amount of code that would be added to quiche-client and quiche-server if we integrated the examples. For instance, I have an a separate branch on top of Datagram that implements WebTransport entirely in the application, see https://github.com/LPardue/quiche/commit/2ed422b57855da4200953ae72736cd9511e8d91e. This adds 400 new lines just for that feature alone. However, those apps are already designed and structured to support multiple Application Layer Protocols, so with a bit of work I'm sure I can fit them in without too much resulting spaghetti.

I'm happy to do the work to "un-feature" Datagram if we think that is the best way to get this landed.

xanathar commented:
Is there anything I can do to help moving this forward?

LPardue commented:
I'm planning to create a commit that de-features Datagram and integrates the example app. But that shouldn't block @ghedo from deeper review of the other library changes.

LPardue commented:
rebased and added a commit to de-feature dgram. Integrating dgram-{client|server} with quiche-{client|server} is a little more work, so I can follow up with that.

LPardue commented:
So I've made https://github.com/cloudflare/quiche/pull/595/commits/78d89242fe8c0948e555b39762568ba301e8fb09. which affects the HTTP/3 API/. The major change is that it ditches `poll_dgram()` and now everything happens under `poll()`. Funnily enough, I spotted some weird code that reminded me that some open questions on the H3 API had been left unanswered, so now was a good time to do that too.

Previously, `poll_dgram` would take an output buffer to write the complete QUIC DATAGRAM too, and then return the flow ID and remaining payload as a copy. This was pretty mad!

Now, `poll` treats Datagrams like streams. It returns a `Datagram` event, and peeks at the datagram to see what the flow ID is (this could in future be used to drop that datagram, skip to a different flow etc. we can do that another day). Then, the new `recv_dgram` function is used to actually get the datagram. Now there is a slight race risk here, because the internal datagram queue could discretionally drop any queued packets, (especially if is is FIFO). I'm not too worried by this,because unreliable apps should expect this kind of behaviour and recv_dgram` returns the atomic DATAGRAM frame payload along with the flow id in that payload. I could add this to the docs, but its a bit verbose.

To support this, some minor tweaks were required to datagram internals.

I also added the C API bindings that have always been missing.

LPardue commented:
In https://github.com/cloudflare/quiche/pull/595/commits/1799489d97e4d9d2a3f03de26816b63ba73b319e I've extracted the salient DATAGRAM behaviour from dgram-{client|server} and now everything is in quiche-{client|server}. I explain what that means below.

I think this addresses all the initial review comments. If folks are happy, I need to squash commits and update the commit comment before merging.

Previously, quiche-client and quiche-server supported HTTP/0.9, HTTP/3 or both for any given QUIC connection. During TLS handshake, ALPN would select one of these and then the applications would create `HttpConn` objects to handle request/response. `Http09Conn` simply invokes quiche transport APIs directly. `Http3Conn` invokes quiche transport and H3 APIs.

With DATAGRAM there are many more permutations. Rather than boil the ocean, this commit introduces support for two modes, with the following CLI parameters that control how datagrams are used:

```
  --dgram-proto PROTO         DATAGRAM application protocol to use.
  --dgram-count COUNT         Number of DATAGRAMs to send.
  --dgram-data DATA           Data to send for certain types of DATAGRAM application protocol.
```

Valid values for `dgram-proto` are `siduck` and `oneway`. 

`siduck` works according to the I-D: the endpoints advertise the ALPN `siduck, siduck-00` and if negotiated, send `dgram_count` number of DATAGRAM frames. The client by default sends `quack`, which can be overridden by the `dgram-data` parameter. The server simply attempts to respond to every received DATAGRAM with `$value-ack`; it ignores `dgram-count` and `dgram-data`. The client counts sent quacks and received quack acks and reports this. No HTTP requests or responses are made in this mode.

`oneway` layers on top of HTTP/3. Endpoitns unilaterally send `dgram-count` amount of HTTP/3 DATAGRAMS to each other and no accounting is done. Clients send on flow ID 0, server on flow ID 1. By default the send `quack`, the server sends `brrr`. Meanwhile, the apps send requests/responses in accordance with all existing CLI parameters.

ghedo commented on src/lib.rs:
```diff
@@ -2973,6 +3073,175 @@ impl Connection {
         self.streams.writable()
     }
 
+    /// Attempts to read the first stored DATAGRAM.
```
Or make it more consistent with the HTTP/3 method docs.

ghedo reviewed (changes requested):
Now that the general comments are addressed I did a more through review, though I might have missed something so I reserve the right to add more comments later :stuck_out_tongue_closed_eyes: 

Also, haven't looked at the apps as I'm not sure I fully understand whats going on there yet, but there are probably enough comments already for now.

ghedo commented on src/lib.rs:
```diff
@@ -452,6 +462,9 @@ impl Config {
             cc_algorithm: CongestionControlAlgorithm::CUBIC,
             hystart: true,
+
+            dgram_recv_max_queue_len: DEFAULT_DGRAM_MAX_QUEUE_LEN,
```
Well, actually, there's nothing being allocated, but it the default len is `0` it might make sense to do `VecDeque::with_capacity()` instead of `VecDeque::new()` in `DatagramQueue::new()`.

LPardue replied:
good catch, this used to be conditional on the feature but that changed

LPardue commented on tools/apps/src/lib.rs:
```diff
@@ -95,27 +99,50 @@ pub struct CommonArgs {
+/// --dgram-proto PROTO         DATAGRAM application protocol.
+/// --dgram-count COUNT         Number of DATAGRAMs to send.
+///  --dgram-data DATA          Data to send for certain types of DATAGRAM
+/// application protocol [default: quack].
```
I think rustfmt did that! Let me check

LPardue replied:
yeah rustmt is a tool. I had to shorten the text

LPardue commented on src/lib.rs:
```diff
@@ -2973,6 +3073,175 @@ impl Connection {
+                // subtract the packet number (max len)
+                max_len = max_len.saturating_sub(packet::MAX_PKT_NUM_LEN);
+                // subtract the crypto overhead
+                max_len = max_len.saturating_sub(frame::MAX_CRYPTO_OVERHEAD);
```
Wor, godo spot. This was fixed in https://github.com/LPardue/quiche/commit/ef26c6462f955a21c7cabca4e9d4cb69b13c67ca. Somehow I've lost that, let me pull it back.

LPardue replied:
fixed in a frankencommit https://github.com/cloudflare/quiche/pull/595/commits/7ae8b247a7a5c0907f87e5889b1e6d2eb9bd579e

xanathar commented on src/lib.rs:
```diff
@@ -2302,6 +2340,41 @@ impl Connection {
+                while let Some(len) = self.dgram_send_queue.peek_front_len() {
+                    if (len + frame::MAX_DGRAM_OVERHEAD) <= left {
+                        // Front of the queue fits this packet, send it
+                        let mut buf = vec![0; len];
```
On the topic of the interface of `dgram_recv` there had been a discussion with some numbers here: https://github.com/LPardue/quiche/pull/1

Ultimately it seemed to boil down to whether to prefer a negligible gain in performance vs a minor uniformity in interface, none of which seemed to be a deal-breaker; we picked one way but the other is equally reasonable.

ghedo replied:
(I clciked resolve by accident...)

> a negligible gain in performance

Did we actually measure this? e.g. removing some copies in https://github.com/cloudflare/quiche/pull/643 increased throughput by a fairly significant margin.

LPardue replied:
An earlier version of this did code did pop and return the `Vec<u8>` directly. This caused a bit of ugliness and inconsistency with the APIs. @xanathar and I had a discussion on `Vec<u8>` (no-copy) vs `&mut [u8]` (copy) on this thread https://github.com/LPardue/quiche/pull/1#issuecomment-614756756

@xanathar provided some benchmarks to compare the two, the numbers seemed pretty marginal for typical MTU-sized DATAGRAM frames.

It's work to change this back to the no-copy model but I think this is a good point for us to agree what we want to do.

ghedo replied:
Even if the external API remained the same, we are still doing an allocation and copy for no reason here. `pop()` can just return the `Vec<u8>` to be used as-is here, and in `dgram_recv()` you can copy it in the output buffer.

LPardue replied:
I've made the non-API change suggested by @ghedo on https://github.com/cloudflare/quiche/pull/595/commits/8333f5bce75861286c951aae4c4679035bfa8c0d

I think this is the best compromise, it really was a bit silly making copies inside quiche when sending.

xanathar commented on src/lib.rs:
```diff
@@ -3632,6 +3907,31 @@ impl Connection {
+                if self
+                    .local_transport_params
+                    .max_datagram_frame_size
+                    .is_none()
```
Not sure I get it, this specific `if` is to address this part of the draft:

> An endpoint that receives a DATAGRAM frame when it has not sent the max_datagram_frame_size transport parameter MUST terminate the connection with error PROTOCOL_VIOLATION.

That's the reason this case exits with `Error::InvalidState`. The full queue state ends up with an `Ok(())` and the head of the queue be discarded.

LPardue replied:
I don't understand the original comment about queues. This is about detecting if the application enabled datagram reception using config/TPs as noted.

I actually hit this while porting the apps into quiche-apps :)

ghedo replied:
So, what I meant basically was that we could check if the queue capacity is `0` instead of the transport param, because I didn't like how this looks :) (the `is_full()` part was just because I got confused on how it works).

But I guess it's the correct check to do, so I'm fine leaving as-is.

LPardue replied:
How about I add a `datagram_enabled()` method that can be used for the check here, then internally it can test TPs or queues as we decide later on?

LPardue replied:
I've done this

xanathar commented on src/lib.rs:
````diff
@@ -2973,6 +3073,175 @@ impl Connection {
+    /// }
+    /// # Ok::<(), quiche::Error>(())
+    /// ```
+    pub fn dgram_max_writable_len(&self) -> Option<usize> {
````
As odd as it seems, `0` is a valid max datagram frame size, and applications might potentially send a zero sized datagram, here I tried to be compliant to that.

I think this was the reference: https://github.com/tfpauly/draft-pauly-quic-datagram/issues/19

In general a lot of the `Option<>` going around are to support this corner case.

LPardue replied:
Yeah its weird but thanks for the reminder of the DATAGRAM draft discussion

xanathar commented on src/lib.rs:
```diff
@@ -732,12 +745,39 @@ impl Config {
+    /// recommended by the current draft of the standard.
+    ///
+    /// The default is `false`.
+    pub fn set_dgram_frames_supported(&mut self, supported: bool) {
```
Fwiw, I really like this suggestion.

LPardue replied:
There's a slight problem with this suggestion because what does `enable_dgram()` mean? Right now we allow an endpoint to opt into a transport feature, and configure some internal queue lengths. 

So what about an application that wants to only send DATAGRAMS and control it's send queue size? It would be counter intuitive to call `enable_dgram(usize, usize)`. I suggest `enable_dgram(bool, usize, usize)`

LPardue commented on src/lib.rs:
```diff
@@ -2489,6 +2562,12 @@ impl Connection {
+        // module which manages retransmission. However, some frames are not
+        // retransmittable and storing them is a waste of resources.
+        // So drop them here.
+        frames.retain(|f| f.retransmittable());
```
@xanathar you added this. Do the alternative sound reasonable?

xanathar replied:
I double-checked the implementation of the recovery and retransmission parts, and setting the vec of the datagram frame to 0 like below, seems to be enough. Haven't tested yet though.

```rust
    // In frame.rs
    pub fn shrink_for_retransmission(&mut self) {
        match &mut self {
            Frame::Datagram { v } => {
                v.clear();
                v.shrink_to_fit();
            },
            _ => (),
        }
    }

    // in lib.rs, this or a for loop doing the same
   frames.iter_mut().for_each(|f| f.shrink_for_retransmission());
```

LPardue replied:
I took @xanathar's snippet, tweaked it slightly, added a test and made https://github.com/cloudflare/quiche/pull/595/commits/a2b87e32639aca3ef6cce2e5ec5e191c0823ee86

in `lib.rs` rather than do a second loop for shedding, I've just put it inside the `for frame in &frames` loop.

LPardue commented on src/lib.rs:
```diff
@@ -306,6 +306,13 @@ const MAX_ACK_RANGES: usize = 68;
 const MAX_STREAM_ID: u64 = 1 << 60;
 
+// The default length of DATAGRAM queues if not specified by the user in config.
+const DEFAULT_DGRAM_MAX_QUEUE_LEN: usize = 1000;
```
I applied this manually

LPardue commented on src/lib.rs:
```diff
@@ -2973,6 +3073,175 @@ impl Connection {
+        self.dgram_send_queue.purge(f);
+    }
+
+    /// Gets the size of the largest Datagram frame payload that can be sent,
```
done

LPardue commented on src/frame.rs:
```diff
@@ -36,6 +36,8 @@ pub const MAX_CRYPTO_OVERHEAD: usize = 8;
 pub const MAX_STREAM_OVERHEAD: usize = 12;
 pub const MAX_STREAM_SIZE: u64 = 1 << 62;
 
+pub const MAX_DGRAM_OVERHEAD: usize = 8;
```
done

LPardue commented on src/h3/ffi.rs:
```diff
@@ -109,6 +109,8 @@ pub extern fn quiche_h3_event_type(ev: &h3::Event) -> u32 {
 
         h3::Event::Finished { .. } => 2,
+
+        h3::Event::Datagram { .. } => 3,
```
done

LPardue commented on src/lib.rs:
```diff
@@ -3256,11 +3525,17 @@ impl Connection {
         }
 
+        #[allow(unused_variables)]
+        let dgram_pending = false;
```
hangover from de-feature. Fixed ta.

LPardue commented on src/lib.rs:
```diff
@@ -3256,11 +3525,17 @@ impl Connection {
+        #[allow(unused_variables)]
+        let dgram_pending = false;
+
+        let dgram_pending = self.dgram_send_queue.has_pending();
```
hangover from de-feature. Fixed ta.

LPardue commented on src/lib.rs:
```diff
@@ -2973,6 +3073,175 @@ impl Connection {
+
+        self.dgram_send_queue.push(buf)?;
+
+        if self.dgram_send_queue.pending_bytes() > self.recovery.cwnd_available()
```
I went with `byte_size` because the term is more bitesized, and makes the code nicer.

LPardue commented on src/lib.rs:
```diff
@@ -4094,6 +4400,15 @@ impl TransportParams {
             b.put_varint(tp.max_ack_delay)?;
         }
 
+        if let Some(max_datagram_frame_size) = tp.max_datagram_frame_size {
```
AKA keep off the grass, but we won't put a sign to tell you. Fixed now.

LPardue commented on src/lib.rs:
```diff
@@ -2973,6 +3073,175 @@ impl Connection {
+    ///
+    /// [`Done`]: enum.Error.html#variant.Done
+    /// [`BufferTooShort`]: enum.Error.html#variant.BufferTooShort
+    pub fn dgram_recv_peek(&self, buf: &mut [u8], len: usize) -> Result<usize> {
```
Naming things is hard.

I'm not precious about this name but the suggestion loses the notion that this is a peek at the receive queue, not write queue. Is that ok? 

Also, it makes this inconsistent with the next method `dgram_recv_front_len` :(

LPardue commented on src/h3/mod.rs:
```diff
@@ -928,6 +1002,25 @@ impl Connection {
+            Ok(_) => {
+                let mut b = octets::Octets::with_slice(&d);
+                let flow_id = b.get_varint()?;
+                Some(flow_id)
```
dome

LPardue commented:
Apologies for the commit spams. I think this addresses nearly all of @ghedo's comments.

One remaining issue is on naming, we can fix that easy. 

One remaining issue that's harder is about `frames.retain(|f| f.retransmittable());` https://github.com/cloudflare/quiche/pull/595#discussion_r500888093. 

Also, the builds are failing due to the nginx patch which now doesn't handle QUICHE_H3_EVENT_DATAGRAM returned by quiche_h3_event_type(). I'm not sure of the ninja-fu that is needed to fix that up.

ghedo commented on src/dgram.rs:
```diff
@@ -0,0 +1,112 @@
+    pub fn discard_front(&mut self) -> Result<()> {
+        match self.queue.pop_front() {
+            None => Err(Error::InvalidState),
+            Some(_) => Ok(()),
```
Should this also update `self.queue_bytes_size`? Could just call `self.pop()` instead now.

ghedo replied:
Also, you can probably simplify this somewhat as:
```rust
self.pop().ok_or(Error::InvalidState)?;
Ok(())
```

ghedo commented on src/frame.rs:
```diff
@@ -33,6 +33,7 @@ use crate::ranges;
 use crate::stream;
 
 pub const MAX_CRYPTO_OVERHEAD: usize = 8;
+pub const MAX_DGRAM_OVERHEAD: usize = 8;
```
Also, FWIW, we are probably never going to use an 8-bytes varint to represent the frame length, so you can probably just use 2 here.

LPardue replied:
agreed, this could in some conditions let us squeeze a few more dgrams in a packet too, so is a net win

ghedo commented on src/frame.rs:
```diff
@@ -687,6 +712,13 @@ impl Frame {
+    pub fn shrink_for_retransmission(&mut self) {
+        if let Frame::Datagram { data } = self {
+            data.clear();
+            data.shrink_to_fit();
```
I think you can just do `*data = Vec::new()` instead of the above two lines.

ghedo commented on src/lib.rs:
```diff
@@ -2287,6 +2316,42 @@ impl Connection {
+                            None => continue,
+                        };
+                    } else if len > max_dgram_payload {
+                        // this dgram frame will never fit. Let's purge it.
```
```suggestion
                        // This dgram frame will never fit. Let's purge it.
```

ghedo commented on src/lib.rs:
```diff
@@ -2446,14 +2511,19 @@ impl Connection {
 
         // Encode frames into the output packet.
-        for frame in &frames {
+        for frame in frames.iter_mut() {
```
```suggestion
        for frame in &mut frames {
```

ghedo commented on src/lib.rs:
```diff
@@ -2958,6 +3048,191 @@ impl Connection {
+    /// [`InvalidState`] is returned if the peer does not support DATAGRAM.
+    /// [`BufferTooShort`] is returned if the DATAGRAM frame length is larger
+    /// than peer's supported DATAGRAM frame length. Use
+    /// `peer_datagram_frame_size` to get the largest supported DATAGRAM
```
Shouldn't this be `dgram_max_writable_len()`? Also, need to add a link to it.

ghedo commented on src/lib.rs:
```diff
@@ -2958,6 +3048,191 @@ impl Connection {
+        let max_payload_len = match self.dgram_max_writable_len() {
+            Some(v) => v as usize,
+            None => {
+                trace!(
```
FWIW, I don;t think we need to log here, since the return code should already be indication of what went wrong.

ghedo commented on src/lib.rs:
```diff
@@ -2958,6 +3048,191 @@ impl Connection {
+        };
+
+        if buf.len() > max_payload_len {
+            trace!("attempt to send DATAGRAM larger than dgram_max_writable_len");
```
Same, regarding no need to log.

ghedo commented on src/lib.rs:
```diff
@@ -3617,6 +3893,27 @@ impl Connection {
+                // frames, as recommended by the standard, so we don't need a
+                // size check.
+                if self.dgram_enabled() {
+                    trace!(
```
Also here. We don't generally log when receiving an invalid frame.

ghedo commented on src/packet.rs:
```diff
@@ -46,7 +46,7 @@ const PKT_NUM_MASK: u8 = 0x03;
 pub const MAX_CID_LEN: u8 = 20;
 
-const MAX_PKT_NUM_LEN: usize = 4;
+pub const MAX_PKT_NUM_LEN: usize = 4;
```
```suggestion
pub const MAX_PKT_NUM_LEN: usize = 4;

```

ghedo commented on src/h3/mod.rs:
```diff
@@ -894,6 +901,60 @@ impl Connection {
         Ok(written)
     }
 
+    /// Sends an HTTP/3 DATAGRAM with the specified flow ID, as defined in
```
```suggestion
    /// Sends an HTTP/3 DATAGRAM with the specified flow ID.
```

ghedo commented on src/h3/mod.rs:
```diff
@@ -894,6 +901,60 @@ impl Connection {
     }
 
+    /// Sends an HTTP/3 DATAGRAM with the specified flow ID, as defined in
+    /// draft-schinazi-quic-h3-datagram-03
```
```suggestion
```

ghedo commented on include/quiche.h:
```diff
@@ -183,6 +183,11 @@ void quiche_config_set_cc_algorithm(quiche_config *config, enum quiche_cc_algori
 // Configures whether to use HyStart++.
 void quiche_config_enable_hystart(quiche_config *config, bool v);
 
+// Enables support for receiving DATAGRAM frames.
```
This needs to be synced with the Rust documentation. Same for many of the other FFI functions below.

ghedo reviewed (changes requested):
Almost there. Added a few more comments for things I missed before, or that came up from the latest changes.

ghedo commented on src/lib.rs:
```diff
@@ -2958,6 +3048,187 @@ impl Connection {
+    /// [`Done`]: enum.Error.html#variant.Done
+    /// [`InvalidState`]: enum.Error.html#variant.InvalidState
+    /// [`BufferTooShort`]: enum.Error.html#variant.BufferTooShort
+    /// [`dgram_max_writable_len()`]: struct.Connection.html#method.on_timeout
```
```suggestion
    /// [`dgram_max_writable_len()`]: struct.Connection.html#method.dgram_max_writable_len
```

ghedo commented on src/h3/mod.rs:
```diff
@@ -507,6 +511,9 @@ pub enum Event {
     /// Stream was closed,
     Finished,
+
+    /// DATAGRAM was received
```
```suggestion
    /// DATAGRAM was received.
```

ghedo reviewed (changes requested):
Two more nits, then it should be good (I can't commit these mysel I think, so need Lucas to press the button).
